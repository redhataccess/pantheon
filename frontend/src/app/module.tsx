import React, { Component } from 'react';
import { Bullseye, Button, Alert, AlertActionCloseButton, FormGroup, TextInput } from '@patternfly/react-core';
import '@app/app.css';
import { Redirect } from 'react-router-dom'

class Module extends Component<any, any> {
  constructor(props) {
    super(props);
    this.state = {
      failedPost: false,
      isMissingFields: false,
      login: false,
      moduleDescription: '',
      moduleFile: File,
      moduleName: '',
      redirect: false
    };
  }

  public componentDidMount() {
    this.checkAuth();
  }

  public render() {
    const { moduleName, moduleDescription, isMissingFields } = this.state;
    return (
      <React.Fragment>
        <Bullseye>
          <div className="app-container">
            <div>
              {isMissingFields && (
                <div className="notification-container">
                  <Alert
                    variant="warning"
                    title="A module name and choosing a file is required."
                    action={<AlertActionCloseButton onClose={this.dismissNotification} />}
                  />
                </div>
              )}

              <FormGroup
                label="Module Name"
                fieldId="module-name"
              >
                <TextInput id="module-name" type="text" placeholder="Module Name" value={moduleName} onChange={this.handleNameInput} />
              </FormGroup>
              <br />
              <FormGroup
                label="Module Description"
                fieldId="module-description"
              >
                <TextInput id="module-description" type="text" placeholder="Module Description" value={moduleDescription} onChange={this.handleModuleInput} />
              </FormGroup>
              <br />
              <FormGroup
                label="Upload .adoc file: "
                fieldId="input-file"
              >
                <input id="input" className="input-file" color="#dddddd" type="file" onChange={this.handleFileChange} />
              </FormGroup>
              <br />
              <Button aria-label="Uploads the .adoc file with the Name and Description specified." onClick={this.saveModule}>Save</Button>
              <div>
                {this.loginRedirect()}
                {this.renderRedirect()}
              </div>
            </div>
          </div>
        </Bullseye>
      </React.Fragment>
    );
  }

  private handleNameInput = moduleName => {
    this.setState({ moduleName });
    // console.log("Name " + moduleName)

  };

  private handleModuleInput = moduleDescription => {
    this.setState({ moduleDescription });
    // console.log("Desc " + moduleDescription)
  };

  private handleFileChange = (event) => {
    this.setFile(event.target.files)
  }

  private setFile = selectorFiles => {
    this.setState({ moduleFile: selectorFiles })
    // console.log(selectorFiles);
  }

  private saveModule = (postBody) => {
    // console.log("My data is: " + this.state.moduleName + " and my desc is " + this.state.moduleDescription + " and my files are " + this.state.moduleFile)
    if (this.state.moduleName === "" || this.state.moduleFile[0] === undefined) {
      this.setState({ isMissingFields: true })
    } else {
      const hdrs = {
        'Accept': 'application/json',
        'cache-control': 'no-cache'
      }
      // console.log("The file is: " + this.state.moduleFile)
      const blob = new Blob([this.state.moduleFile[0]])
      const formData = new FormData();
      formData.append("jcr:title", this.state.moduleName)
      formData.append("jcr:description", this.state.moduleDescription)
      formData.append("sling:resourceType", "pantheon/modules")
      formData.append("jcr:primaryType", 'pant:module')
      formData.append("asciidoc@TypeHint", 'nt:file')
      formData.append("asciidoc/jcr:content/jcr:mimeType", "text/x-asciidoc")
      formData.append("asciidoc", blob)
      formData.append(":operation", "pant:newModuleVersion")

      fetch('/content/modules/' + this.state.moduleName, {
        body: formData,
        headers: hdrs,
        method: 'post'
      }).then(response => {
        if (response.status === 201 || response.status === 200) {
          console.log(" Works " + response.status)
          this.setState({ redirect: true })
        } else if (response.status === 500) {
          console.log(" Needs login " + response.status)
          this.setState({ login: true })
        } else {
          console.log(" Failed " + response.status)
          this.setState({ failedPost: true })
        }
      });
    }
  }

  private renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/' />
    } else {
      return ""
    }
  }

  private loginRedirect = () => {
    if (this.state.login) {
      return <Redirect to='/login' />
    } else {
      return ""
    }
  }

  private checkAuth = () => {
    fetch("/system/sling/info.sessionInfo.json")
      .then(response => response.json())
      .then(responseJSON => {
        const key = "userID"
        if (responseJSON[key] === 'anonymous') {
          this.setState({ login: true })
        }
      })
  }

  private dismissNotification = () => {
    this.setState({ isMissingFields: false });
  };

}

export { Module }