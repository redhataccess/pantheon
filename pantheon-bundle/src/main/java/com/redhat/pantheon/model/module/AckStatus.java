package com.redhat.pantheon.model.module;

import java.util.Calendar;

import javax.inject.Named;

import org.apache.jackrabbit.JcrConstants;

import com.redhat.pantheon.model.api.Field;
import com.redhat.pantheon.model.api.WorkspaceChild;
import com.redhat.pantheon.model.api.annotation.JcrPrimaryType;

@JcrPrimaryType("pant:acknowledgment")
public interface AckStatus extends WorkspaceChild {
    
    @Named("pant:message")
    Field<String> message();
    
    @Named("pant:status")
    Field<String> status();
    
    @Named("pant:sender")
    Field<String> sender();
    
    @Named("jcr:lastModifiedBy")
    Field<String> lastModifiedBy();

    @Named(JcrConstants.JCR_LASTMODIFIED)
    Field<Calendar> dateModified();
}
