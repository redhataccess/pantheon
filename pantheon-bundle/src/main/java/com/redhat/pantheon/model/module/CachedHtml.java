package com.redhat.pantheon.model.module;

import javax.inject.Named;

import com.redhat.pantheon.model.api.Field;
import com.redhat.pantheon.model.api.WorkspaceChild;

/**
 * Models a single instance of a module's cachedHtml content.
 */
public interface CachedHtml extends WorkspaceChild {

    @Named("pant:hash")
    Field<String> hash();

    @Named("jcr:data")
    Field<String> data();
}