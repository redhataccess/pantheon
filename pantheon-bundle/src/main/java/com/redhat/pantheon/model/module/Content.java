package com.redhat.pantheon.model.module;

import com.redhat.pantheon.model.api.Child;
import com.redhat.pantheon.model.api.Field;
import com.redhat.pantheon.model.api.FileResource;
import com.redhat.pantheon.model.api.SlingResource;
import org.apache.sling.api.resource.Resource;

/**
 * Models a single instance of a module's content. Multiple content instances may be found on a
 * given module representing several content versions.
 */
public class Content extends SlingResource {

    public final Field<String> asciidocContent = stringField("asciidoc/jcr:content/jcr:data");

    public final Child<CachedContent> cachedHtml = child("cachedHtml", CachedContent.class);

    public final Child<FileResource> asciidoc = file("asciidoc");

    public Content(Resource wrapped) {
        super(wrapped);
    }

    /**
     * A child resource for a {@link ModuleVersion} which contains cached data
     * when a resource is generated.
     */
    public static class CachedContent extends SlingResource {

        public final Field<String> hash = stringField("pant:hash");

        public final Field<String> data = stringField("jcr:data");

        public CachedContent(Resource wrapped) {
            super(wrapped);
        }
    }
}
