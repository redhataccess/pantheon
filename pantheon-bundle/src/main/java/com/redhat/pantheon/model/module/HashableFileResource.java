package com.redhat.pantheon.model.module;

import com.redhat.pantheon.model.api.Field;
import com.redhat.pantheon.model.api.FileResource;
import com.redhat.pantheon.model.api.annotation.JcrMixins;
import com.redhat.pantheon.model.api.annotation.JcrPrimaryType;

import javax.inject.Named;

/**
 * Extension of the {@link FileResource} which adds a hash field.
 * @author Carlos Munoz
 */
@JcrPrimaryType("nt:file")
@JcrMixins("pant:hashable")
public interface HashableFileResource extends FileResource {

    @Named("pant:hash")
    Field<String> hash();
}
