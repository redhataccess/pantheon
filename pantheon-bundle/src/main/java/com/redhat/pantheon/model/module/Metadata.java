package com.redhat.pantheon.model.module;

import com.redhat.pantheon.model.api.Field;
import com.redhat.pantheon.model.api.SlingResource;
import org.apache.jackrabbit.JcrConstants;
import org.apache.sling.api.resource.Resource;

import java.util.Calendar;

/**
 * Models an instance of metadata for a module. Multiple metadata
 * instances may found on a given module representing different
 * versions of said metadata.
 */
public class Metadata extends SlingResource {

    public final Field<String> title = stringField("jcr:title");

    public final Field<String> mAbstract = stringField("pant:abstract");

    public final Field<String> headline = stringField("pant:headline");

    public final Field<String> description = stringField("jcr:description");

    public final Field<Calendar> created = dateField("jcr:created");

    public final Field<String> createdBy = stringField("jcr:createdBy");

    public final Field<ModuleType> moduleType = enumField("pant:moduleType", ModuleType.class);

    public final Field<Calendar> datePublished = dateField("pant:datePublished");

    public final Field<Calendar> dateUploaded = dateField("pant:dateUploaded");

    public final Field<Calendar> dateModified = dateField(JcrConstants.JCR_LASTMODIFIED);

    public final Field<String> primaryType = stringField("jcr:primaryType");

    public Metadata(Resource wrapped) {
        super(wrapped);
    }

}
