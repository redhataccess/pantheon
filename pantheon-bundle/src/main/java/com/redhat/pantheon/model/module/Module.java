package com.redhat.pantheon.model.module;

import com.redhat.pantheon.model.api.SlingResource;
import com.redhat.pantheon.model.api.annotation.JcrPrimaryType;
import org.apache.sling.api.resource.Resource;

import javax.annotation.Nonnull;
import javax.jcr.RepositoryException;
import java.util.Locale;
import java.util.Optional;

import static java.util.Optional.empty;
import static java.util.Optional.ofNullable;

/**
 * The definition of a Module resource in the system.
 * Module's contains different versions for different languages.
 * <br/><br/>
 *
 * A module's structure in the JCR tree is as follows:
 * .../modulename
 *                       /en-US
 *                             /4
 *                                   /content
 *                                           /asciidoc
 *                                           /cachedHtml
 *                                   /metadata
 *                             /3
 *                                   /content
 *                                   /metadata
 *                             /2 (older - just for historical purposes)
 *                                   /content
 *                                   /metadata
 *                             /1
 *                                   /content
 *                                   /metadata
 */
@JcrPrimaryType("pant:module")
public class Module extends SlingResource {

    public Module(@Nonnull Resource resource) {
        super(resource);
    }

    public ModuleLocale getModuleLocale(Locale locale) {
        return child(locale.toString(), ModuleLocale.class).get();
    }

    public ModuleLocale getOrCreateModuleLocale(Locale locale) {
        return child(locale.toString(), ModuleLocale.class).getOrCreate();
    }

    public ModuleLocale createModuleLocale(Locale locale) {
        return child(locale.toString(), ModuleLocale.class).create();
    }

    public Optional<ModuleVersion> getDraftVersion(@Nonnull final Locale locale) {
        ModuleLocale moduleLocale = getModuleLocale(locale);
        if(moduleLocale != null) {
            try {
                return ofNullable( moduleLocale.draft.getReference() );
            } catch (RepositoryException e) {
                throw new RuntimeException(e);
            }
        }
        return empty();
    }

    public Optional<ModuleVersion> getReleasedVersion(@Nonnull final Locale locale) {
        ModuleLocale moduleLocale = getModuleLocale(locale);
        if(moduleLocale != null) {
            try {
                return ofNullable( moduleLocale.released.getReference() );
            } catch (RepositoryException e) {
                throw new RuntimeException(e);
            }
        }
        return empty();
    }

    /**
     * @param locale The locale to fetch the content instance for.
     * @return The released content for a given locale
     */
    public Optional<Content> getReleasedContent(final Locale locale) {
        return getReleasedVersion(locale)
                .map(moduleVersion -> moduleVersion.content.get());
    }

    /**
     * @param locale The locale to fetch the content instance for.
     * @return The draft content for a given locale
     */
    public Optional<Content> getDraftContent(final Locale locale) {
        return getDraftVersion(locale)
                .map(moduleVersion -> moduleVersion.content.get());
    }

    /**
     * @param locale The locale to fetch the content instance for.
     * @return The released metadata for a given locale
     */
    public Optional<Metadata> getReleasedMetadata(final Locale locale) {
        return getReleasedVersion(locale)
                .map(moduleVersion -> moduleVersion.metadata.get());
    }

    /**
     * @param locale The locale to fetch the content instance for.
     * @return The draft metadata for a given locale
     */
    public Optional<Metadata> getDraftMetadata(final Locale locale) {
        return getDraftVersion(locale)
                .map(moduleVersion -> moduleVersion.metadata.get());
    }
}
