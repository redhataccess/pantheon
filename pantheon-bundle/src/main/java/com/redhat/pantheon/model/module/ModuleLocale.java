package com.redhat.pantheon.model.module;

import com.redhat.pantheon.model.api.ReferenceField;
import com.redhat.pantheon.model.api.SlingResource;
import com.redhat.pantheon.model.api.annotation.JcrPrimaryType;
import org.apache.sling.api.resource.Resource;

import javax.annotation.Nonnull;

import static com.google.common.collect.Streams.stream;
import static java.util.stream.Collectors.counting;

/**
 * A specific module locale node which houses all the versions for a specific language in the module.
 */
@JcrPrimaryType("pant:moduleLocale")
public class ModuleLocale extends SlingResource {

    public final ReferenceField<ModuleVersion> released = referenceField("released", ModuleVersion.class);

    public final ReferenceField<ModuleVersion> draft = referenceField("draft", ModuleVersion.class);

    public ModuleLocale(@Nonnull Resource resource) {
        super(resource);
    }

    public ModuleVersion getVersion(String name) {
        return child(name, ModuleVersion.class).get();
    }

    public ModuleVersion getOrCreateVersion(String name) {
        return child(name, ModuleVersion.class).getOrCreate();
    }

    public ModuleVersion createNextVersion() {
        // Generate a new version name
        return child(generateNextVersionName(), ModuleVersion.class).create();
    }

    private String generateNextVersionName() {
        return "" + (stream(this.getChildren()).collect(counting()) + 1);
    }
}
