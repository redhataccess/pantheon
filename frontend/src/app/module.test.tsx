import React from 'react';
import { Module } from '@app/module';
import "isomorphic-fetch"

import { mount, shallow } from 'enzyme';
import { Bullseye, TextInput, FormGroup, Button } from '@patternfly/react-core';

import sinon from "sinon";

describe('Login tests', () => {
  test('should render Login component', () => {
    const view = shallow(<Module />);
    expect(view).toMatchSnapshot();
  });

  it('should render a Bullseye layout', () => {
    const wrapper = mount(<Module />);
    const bullseyeLayout = wrapper.find(Bullseye);
    expect(bullseyeLayout.exists()).toBe(true)
  });


  it('should render a form group', () => {
    const wrapper = mount(<Module />);
    const formGroup = wrapper.find(FormGroup);
    expect(formGroup.exists()).toBe(true)
  });

  it('should render a text input', () => {
    const wrapper = mount(<Module />);
    const textInput = wrapper.find(TextInput);
    expect(textInput.exists()).toBe(true)
  });

  it('should render a Button', () => {
    const wrapper = mount(<Module />);
    const button = wrapper.find(Button);
    expect(button.exists()).toBe(true)
  });

  it('test click event', () => {
    const mockCallBack = jest.fn();
  
    const button = shallow((<Button onClick={mockCallBack}>Submit</Button>));
    button.find('button').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  test('saveModule() click event', () => {
    const wrapper = shallow(<Module />);
    const instance = wrapper.instance();
    const spy = sinon.spy(instance, 'saveModule');

    wrapper.setState({buildDate: '',commitHash: '', commitText: ''})
    wrapper.find(Button).at(0).simulate('click');
    sinon.assert.calledOnce(spy);
  });

});
