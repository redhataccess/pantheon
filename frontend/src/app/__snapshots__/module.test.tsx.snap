// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login tests should render Login component 1`] = `
<Fragment>
  <Bullseye>
    <div
      className="app-container"
    >
      <div>
        <FormGroup
          fieldId="module-name"
          label="Module Name"
        >
          <TextInput
            aria-label={null}
            className=""
            id="module-name"
            isDisabled={false}
            isReadOnly={false}
            isRequired={false}
            isValid={true}
            onChange={[Function]}
            placeholder="Module Name"
            type="text"
            value=""
          />
        </FormGroup>
        <br />
        <FormGroup
          fieldId="module-description"
          label="Module Description"
        >
          <TextInput
            aria-label={null}
            className=""
            id="module-description"
            isDisabled={false}
            isReadOnly={false}
            isRequired={false}
            isValid={true}
            onChange={[Function]}
            placeholder="Module Description"
            type="text"
            value=""
          />
        </FormGroup>
        <br />
        <FormGroup
          fieldId="input-file"
          label="Upload .adoc file: "
        >
          <input
            className="input-file"
            color="#dddddd"
            id="input"
            onChange={[Function]}
            type="file"
          />
        </FormGroup>
        <br />
        <Button
          aria-label="Uploads the .adoc file with the Name and Description specified."
          onClick={[Function]}
        >
          Save
        </Button>
        <div />
      </div>
    </div>
  </Bullseye>
</Fragment>
`;

exports[`Login tests test checkAuth function 1`] = `undefined`;

exports[`Login tests test dismissNotification function 1`] = `undefined`;

exports[`Login tests test handleFileChange function 1`] = `undefined`;

exports[`Login tests test handleModuleInput function 1`] = `undefined`;

exports[`Login tests test handleNameInput function 1`] = `undefined`;

exports[`Login tests test loginRedirect function 1`] = `""`;

exports[`Login tests test renderRedirect function 1`] = `""`;

exports[`Login tests test setFile function 1`] = `undefined`;
